Use same plugins as in repo.continuum.io/pkgs/free

mbargull opened the pull request:
ref: https://github.com/conda-forge/nsis-feedstock/pull/9

mbargull committed:
fix errorlevel test

mbargull committed:
install into NSIS subfolder

mbargull committed:
use same plugins as used for repo.continuum.io/pkgs/free

mbargull committed:
add comment for hierarchy reduction in unzip

jakirkham commented on recipe/bld.bat:
```diff
@@ -1,22 +1,42 @@
-robocopy . "%PREFIX%" /S /XF bld.bat
-if errorlevel NEQ 1 exit 1
+set "PREFIX_NSIS=%PREFIX%\NSIS"
+robocopy . "%PREFIX_NSIS%" /S /XF bld.bat Docs Examples
```
So we had been dumping things into `%PREFIX%` directly. Have no problem changing it for 3.x. How does `conda` make sure this stuff gets added to the `%PATH%` though?

mbargull replied:
Just mirrored the behavior from the Anaconda recipe.
This is simply what `constructor` expects, see: https://github.com/conda/constructor/blob/b2891730f774b97e646a80476656e29eaca2193a/constructor/winexe.py#L26
Well, with this is won't be on `PATH` 😕 ... What do think would be the best way to also make sure it's accessible via `PATH`? I think dumping all its subfolders into one dedicated `NSIS` folder is a good thing, but having its `.exe` files available on `PATH` would be nice, too.

jakirkham replied:
Can think of a couple of solution.

One would be an activate script that adds these directories to the path. This is probably the most effective and activate scripts are pretty reliable. Though it requires that people actually activate the environment for them to work.

Another would be placing the executables in both locations. Not sure how library loading proceeds here for the executables should they need them.

mbargull replied:
Placing them in other locations won't work since NSIS requires that directory layout. I was thinking if symlinks were possible, but having thought about it, I don't think we should use those, esp. on Windows.
Activate script it is! (And overall we shouldn't really care if people don't activate the environments beforehand since that's just not how Conda should be used :wink:)

jakirkham commented on recipe/bld.bat:
```diff
@@ -1,22 +1,42 @@
-if errorlevel NEQ 1 exit 1
+set "PREFIX_NSIS=%PREFIX%\NSIS"
+robocopy . "%PREFIX_NSIS%" /S /XF bld.bat Docs Examples
+if errorlevel 1 exit 1
```
Yeah I have no idea why we had `NEQ` before. Was worried it was some bizarro `robocopy` workaround. Thanks for fixing it.

jakirkham commented on recipe/bld.bat:
```diff
@@ -1,22 +1,42 @@
 if errorlevel 1 exit 1
 
-copy "UAC\Plugins\x86-unicode\UAC.dll " "%PREFIX%\Plugins\x86-unicode\"
+copy "elevate\bin.x86-32\elevate.exe" "%PREFIX_NSIS%\Plugins\x86-unicode\"
```
Do you know why this is installed in `Plugins`? Seems like it should go in `Bin` being an `exe` and all. 😕

jakirkham replied:
FWIW I know this is done in the Anaconda recipe. Just trying to understand the motivation for this choice.

mbargull replied:
No idea, maybe @mingwandroid can enlighten us 😉

jakirkham commented on recipe/bld.bat:
```diff
@@ -1,22 +1,42 @@
+copy "UAC\U\UAC.dll" "%PREFIX_NSIS%\Plugins\x86-unicode\"
 if errorlevel 1 exit 1
-copy "UnicodePathTest_1.0\Plugin\UnicodePathTest.dll" "%PREFIX%\Plugins\x86-unicode\"
+copy "untgz\Plugins\x86-unicode\untgz.dll" "%PREFIX_NSIS%\Plugins\x86-unicode\"
```
Guessing this falls out of changes to the decompression step like respecting the root directory and such.

mbargull replied:
Sorry, I don't quite understand your comment. ("non-native speaker" excuse :wink:)

jakirkham replied:
That's probably because I was rambling. 😄 Feel free to ignore this comment.

jakirkham commented on recipe/bld.bat:
```diff
@@ -1,22 +1,42 @@
+
+echo(||exit 1
+echo(=== elevate.exe license information ===||exit 1
+echo(UKNOWN LICENSE||exit 1
```
`UKNOWN` -> `UNKNOWN`

jakirkham commented on recipe/bld.bat:
```diff
@@ -1,22 +1,42 @@
+
+echo(||exit 1
+echo(=== UnicodePathTest plugin license information ===||exit 1
+type "UnicodePathTest\Readme.txt"||exit 1
```
Yeah I guess that works. They at least say LGPL. Still its not a lot to go on. Thanks for adding it though.

mbargull replied:
Ah, yeah, forgot to comment on that. But yes, you got it right, this was the only place I could find that mentioned a license. Not optimal but better than nothing, I suppose.

jakirkham commented on recipe/bld.bat:
```diff
@@ -1,22 +1,42 @@
+echo(||exit 1
+echo(=== elevate.exe license information ===||exit 1
+echo(UKNOWN LICENSE||exit 1
+) > "LICENSE.txt"||exit 1
```
Nice! This is much cleaner thanks for updating this. 😄

jakirkham commented on recipe/get_plugins.py:
```diff
@@ -47,42 +49,96 @@ def checksum(htname, fname):
+                        members.append(zip_info)
+                fh.extractall(dname, members)
+            else:
+                fh.extractall(dname)
```
So when I wrote this I thought, "Hey! We can unzip stuff in Python. :) Plus I have played with `ZipFile` a lot lately. This should be easy." Looking back on it, especially with the `7za` shell out, am wondering if we wouldn't be better off using `7za` for all cases. Thoughts?

mbargull replied:
I only added that part so we don't have to write the commit checksum from the GitHub archives multiple times in `get_plugins.py` and `bld.bat` 😉 .
I just tried `7za` with the additional parameter
* `-spe` : eliminate duplication of root folder for extract command
but that didn't do anything... 😕

jakirkham replied:
So `7za` won't work on the zip files?

mbargull replied:
`7za` could extract the zip files but I don't see how we could tell it to skip creating the topmost directory. Currently I think that Python snippet is the most convenient solution.

jakirkham commented on recipe/get_plugins.py:
```diff
@@ -47,42 +49,96 @@ def checksum(htname, fname):
+        dname="elevate"
+    ),
+    Download(
+        url="http://nsis.sourceforge.net/mediawiki/images/7/79/UAC_v0.2.2d.zip",
```
Guessing this is the version Anaconda is using? Seems fine to me if so.

mbargull replied:
Yes, that's the archive I found which had the `dll` with the same checksum as in Anaconda.

jakirkham commented on recipe/get_plugins.py:
```diff
@@ -47,42 +49,96 @@ def checksum(htname, fname):
     ),
     Download(
-        url="http://nsis.sourceforge.net/mediawiki/images/a/a7/UnicodePathTest_1.0.zip",
+        url="https://github.com/mingwandroid/nsis-untgz/archive/5c814c5f2c8a9a14e3a6ddd3e594fcc76db5b86a.zip",
```
Any idea why the plugin url is not used?

mbargull replied:
Again, just to get the file with the same checksum as in the Anaconda recipe.
Maybe @mingwandroid can tell you how the plugin was altered.

jakirkham commented on recipe/get_plugins.py:
```diff
@@ -47,42 +49,96 @@ def checksum(htname, fname):
     ),
     Download(
-        url="http://nsis.sourceforge.net/mediawiki/images/9/9d/Untgz.zip",
+        url="https://github.com/mingwandroid/nsis-UnicodePathTest/archive/fa74caef553883f1820049d89e169aff57551796.zip",
```
Any idea why the plugin url is not used?

mbargull replied:
Again, just to get the file with the same checksum as in the Anaconda recipe.
Maybe @mingwandroid can tell you how the plugin was altered.

jakirkham commented on recipe/get_plugins.py:
```diff
@@ -47,42 +49,96 @@ def checksum(htname, fname):
+]
+
+for hashed_file in files:
+    check_hash(hashed_file, hashed_file.fname)
```
Oh! I guess we could do this. Are we worried that these will be changed that often/we will be deviating that much?

mbargull replied:
No idea. @mingwandroid (@ Ray: sorry for that notification spree 😉) to the help 😄

mbargull replied:
(And of course those checks are also redundant since the downloads are already checksum-checked. They are more like initial sanity checks/documentation for this PR.)

jakirkham commented on recipe/meta.yaml:
```diff
@@ -16,11 +16,12 @@ build:
   home: http://sourceforge.net/projects/nsis
   summary: Nullsoft Scriptable Install System
-  license: CPL-1.0, zlib, bzip2
+  license: CPL-1.0, zlib, bzip2, UNKNOWN
```
We should probably add LGPL. I think there was some kind of BSD as well, but am not remembering.

mbargull replied:
You are right. Will do.

jakirkham commented:
Thanks for doing this @mbargull! This definitely looks much better than what I had. Added some comments and questions above. Please let me know your thoughts. :)

mbargull committed:
fix/add more licenses

mbargull committed:
add activate script

mbargull commented:
Added an activate script. Are there any conventions on naming those scripts?
I also changed the test for the `errorlevel` ~since I don't know why it only tested for 2 or higher.~
EDIT: Ok, `makensis /?` returns an error code `1`, so that test was actually correct. Nonetheless I changed the command to `makensis /VERSION` which returns `0` instead.

Best thing would be to merge this into your PR so that we can see if the CI/Windows is happy with those changes.

jakirkham commented on recipe/nsis_activate.bat:
```diff
@@ -0,0 +1 @@
+@SET "PATH=%CONDA_PREFIX%\NSIS;%PATH%"
```
We should also have a deactivate script. This may require storing a backup of the path during activate.

jakirkham replied:
Also we normally just call them `activate.bat` and `deactivate.bat` as the `nsis` is already clear by being part of this recipe.

mbargull replied:
> We should also have a deactivate script.

Yes! That was a brain fart on my end here.

> just call them `activate.bat` and `deactivate.bat`

For the recipe itself I think this is a good call. But then I should change the filename when copying in `bld.bat`, right? Skimming the code for `conda`/`conda-build` I didn't see any indication those scripts will get renamed by them. So if not named acc. to the package, they might clobber de-/activate scripts from other packages.

jakirkham replied:
Correct. Pretty much most activate/deactivate script installs mimic the [`toolchain` recipe]( https://github.com/conda-forge/toolchain-feedstock ). Here's the [Windows build script for the `toolchain`]( https://github.com/conda-forge/toolchain-feedstock/blob/f87689631ac261cb705b9ade2553edf439a56d9f/recipe/bld.bat ). One could probably be more clever and substitute `toolchain` in that recipe with `%PKG_NAME%`.

jakirkham commented:
Minor comments on the activate script. Once they are addressed would be happy to merge.

jakirkham commented:
Do you have time to do the last few things @mbargull? If not, I can just merge as is and clean them up as time allows. Either way.

mbargull committed:
add/improve (de-)activate scripts

mbargull commented:
Sorry for the delay! Slipped through on my ToDo list..

mbargull commented:
I just noticed that http://code.kliu.org/misc/elevate/elevate-1.3.0-redist.7z doesn't seem to be available anymore. Still available at the Internet Archive: http://web.archive.org/web/20161025050009if_/http://code.kliu.org/misc/elevate/elevate-1.3.0-redist.7z. What's the preferred way to deal with this?

jakirkham commented:
Ouch! Maybe bundle it in the recipe for now. Alternatively we can add it to tag on the repo. The latter is what we typically do. Am happy to do that.

mbargull commented:
Ok, I'll leave that decision up to you. You'll know best, what to do :D.
Did a quick Google search for the filename/URL and it came up with this:
* Source code-only mirror:
  https://github.com/jviotti/elevate/tree/6abaa3074d5f56aa6b7e3d4cdb928fd2d995f61f
* Source code and binary mirror:
  https://github.com/markhuber/modern-ie-vagrant-builder/tree/aaabd48377ed8813428f3500c233ff7e023b323d/bin
  * direct link to `elevate-1.3.0-redist.7z`: https://github.com/markhuber/modern-ie-vagrant-builder/raw/aaabd48377ed8813428f3500c233ff7e023b323d/bin/elevate-1.3.0-redist.7z
  * direct link to `elevate.exe`: https://github.com/markhuber/modern-ie-vagrant-builder/raw/aaabd48377ed8813428f3500c233ff7e023b323d/bin/elevate/elevate-1.3.0-redist.7z%20Folder/bin.x86-32/elevate.exe

If we use the direct URL for `elevate.exe`, we can also leave out the `7za` build dependency. Though I don't know whether there would be any downsides in using some random `https://github.com/.../raw/...` URL. But, as I said, I'll leave that up to you :wink:. From my side this is ready to be merged into your PR to the main repo.

jakirkham commented:
Sorry missed your comment here. My bad! 😞 Thank for all the links and info.

That said, was just thinking the other day that this should already be merged. Let's sort out the details in the main PR. Thanks for all of your hard work here @mbargull! 😄

jakirkham commented:
Also weird thing. Seems like http://code.kliu.org/misc/elevate/elevate-1.3.0-redist.7z is back. 😕 Maybe there were some intermittent issues?

jakirkham commented on recipe/bld.bat:
```diff
@@ -1,22 +1,49 @@
+echo(||exit 1
+echo(=== elevate.exe license information ===||exit 1
+echo(UNKNOWN LICENSE||exit 1
+) > "LICENSE.txt"||exit 1
```
Side note: If you haven't seen issue ( https://github.com/conda/conda-build/issues/1247 ), it might be worth taking a look.
